File: Sources/MoonParser/Patterns.swift
import MoonAST
import MoonLexer

struct ParsePatternOptions {
    var allowApp: Bool = true
}

func parsePattern(_ ts: TokenStream, options: ParsePatternOptions = ParsePatternOptions()) throws -> Pattern {
    if options.allowApp == false {
        return try parsePatternAtom(ts)
    }
    return try parsePatternApp(ts)
}

private func parsePatternApp(_ ts: TokenStream) throws -> Pattern {
    let start = ts.peek()
    let pat = try parsePatternAtom(ts)

    switch pat {
    case .pVar, .pCon:
        var args: [Pattern] = []
        while isPatternAtomStart(ts) {
            args.append(try parsePatternAtom(ts))
        }
        if !args.isEmpty {
            let name: String
            switch pat {
            case .pVar(let n, _): name = n
            case .pCon(let n, _, _): name = n
            default: name = ""
            }
            return .pCon(name: name, args: args, span: ts.makeSpan(start: start, end: ts.last()))
        }
    default:
        break
    }

    return pat
}

private func isPatternAtomStart(_ ts: TokenStream) -> Bool {
    ts.check(.ident, .lparen, .lbracket, .string, .int, .float, .kwTrue, .kwFalse)
}

private func parsePatternAtom(_ ts: TokenStream) throws -> Pattern {
    let start = ts.peek()

    if ts.at(.lparen) {
        _ = ts.advance()
        var elements: [Pattern] = []
        if !ts.at(.rparen) {
            elements.append(try parsePattern(ts))
            while ts.skip(.comma) {
                elements.append(try parsePattern(ts))
            }
        }
        let end = try ts.expect(.rparen)
        return .pTuple(elements: elements, span: ts.makeSpan(start: start, end: end))
    }

    if ts.at(.lbracket) {
        _ = ts.advance()
        var elements: [Pattern] = []
        if !ts.at(.rbracket) {
            elements.append(try parsePattern(ts))
            while ts.skip(.comma) {
                elements.append(try parsePattern(ts))
            }
        }
        let end = try ts.expect(.rbracket)
        return .pList(elements: elements, span: ts.makeSpan(start: start, end: end))
    }

    if ts.check(.string, .int, .float, .kwTrue, .kwFalse) {
        let lit = try parseLiteral(ts)
        return .pLit(value: lit, span: literalSpan(lit))
    }

    if ts.at(.ident) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        let value = tok.value ?? ""
        if value == "_" {
            return .pWildcard(span: span)
        }
        if let first = value.first, first.isUppercase {
            return .pCon(name: value, args: [], span: span)
        }
        return .pVar(name: value, span: span)
    }

    throw ParseError("Expected pattern", line: start.line, column: start.column)
}

File: legacy/packages/parser/src/patterns.ts
import type { Pattern } from "@moon/ast";
import { ParseError, type TokenStream } from "./token-stream";
import { parseLiteral } from "./literals";

export interface ParsePatternOptions {
  allowApp?: boolean;
}

export function parsePattern(ts: TokenStream, opts: ParsePatternOptions = {}): Pattern {
  if (opts.allowApp === false) {
    return parsePatternAtom(ts);
  }
  return parsePatternApp(ts);
}

function parsePatternApp(ts: TokenStream): Pattern {
  const start = ts.peek();
  let pat = parsePatternAtom(ts);

  if (pat.kind === "PVar" || pat.kind === "PCon") {
    const args: Pattern[] = [];
    while (isPatternAtomStart(ts)) {
      args.push(parsePatternAtom(ts));
    }
    if (args.length > 0) {
      const end = ts.last();
      return { kind: "PCon", name: pat.name, args, span: ts.makeSpan(start, end) };
    }
  }

  return pat;
}

function isPatternAtomStart(ts: TokenStream): boolean {
  return ts.check("IDENT", "LPAREN", "LBRACKET", "STRING", "INT", "FLOAT", "KW_TRUE", "KW_FALSE");
}

function parsePatternAtom(ts: TokenStream): Pattern {
  const start = ts.peek();

  if (ts.at("LPAREN")) {
    ts.advance();
    const elements: Pattern[] = [];
    if (!ts.at("RPAREN")) {
      elements.push(parsePattern(ts));
      while (ts.skip("COMMA")) {
        elements.push(parsePattern(ts));
      }
    }
    const end = ts.expect("RPAREN");
    return { kind: "PTuple", elements, span: ts.makeSpan(start, end) };
  }

  if (ts.at("LBRACKET")) {
    ts.advance();
    const elements: Pattern[] = [];
    if (!ts.at("RBRACKET")) {
      elements.push(parsePattern(ts));
      while (ts.skip("COMMA")) {
        elements.push(parsePattern(ts));
      }
    }
    const end = ts.expect("RBRACKET");
    return { kind: "PList", elements, span: ts.makeSpan(start, end) };
  }

  if (ts.check("STRING", "INT", "FLOAT", "KW_TRUE", "KW_FALSE")) {
    const lit = parseLiteral(ts);
    return { kind: "PLit", value: lit, span: lit.span };
  }

  if (ts.at("IDENT")) {
    const tok = ts.advance();
    if (tok.value === "_") {
      return { kind: "PWildcard", span: ts.makeSpan(start, tok) };
    }
    if (/^[A-Z]/.test(tok.value!)) {
      return { kind: "PCon", name: tok.value!, args: [], span: ts.makeSpan(start, tok) };
    }
    return { kind: "PVar", name: tok.value!, span: ts.makeSpan(start, tok) };
  }

  throw new ParseError("Expected pattern", start.line, start.column);
}